drivers/charger/charger_bq25180.c

```diff
@@ -134,10 +140,78 @@ static int bq25180_get_charge_current(const struct device *dev,
 	return 0;
 }
 
+static int bq25180_get_online(const struct device *dev,
+			      enum charger_online *online)
+{
+	const struct bq25180_config *cfg = dev->config;
+	uint8_t val;
+	int ret;
+
+	ret = i2c_reg_read_byte_dt(&cfg->i2c, BQ25180_STAT0, &val);
+	if (ret < 0) {
+		return ret;
+	}
+
+	if ((val & BQ25180_STAT0_VIN_PGOOD_STAT) != 0x00) {
+		*online = CHARGER_ONLINE_FIXED;
+	} else {
+		*online = CHARGER_ONLINE_OFFLINE;
+	}
+
+	return 0;
+}
+
+static int bq25180_get_status(const struct device *dev,
+			      enum charger_status *status)
+{
+	const struct bq25180_config *cfg = dev->config;
+	uint8_t val;
+	int ret;
+
+	ret = i2c_reg_read_byte_dt(&cfg->i2c, BQ25180_ICHG_CTRL, &val);
+	if (ret < 0) {
+		return ret;
+	}
+
+	if ((val & BQ25180_ICHG_CHG_DIS) != 0x00) {
+		*status = CHARGER_STATUS_NOT_CHARGING;
+		return 0;
+	}
+
+	ret = i2c_reg_read_byte_dt(&cfg->i2c, BQ25180_STAT0, &val);
+	if (ret < 0) {
+		return ret;
+	}
+
+	if ((val & BQ25180_STAT0_VIN_PGOOD_STAT) == 0x00) {
+		*status = CHARGER_STATUS_DISCHARGING;
+		return 0;
+	}
```

Comment:
```suggestion
	ret = i2c_reg_read_byte_dt(&cfg->i2c, BQ25180_STAT0, &val);
	if (ret < 0) {
		return ret;
	}

	if ((val & BQ25180_STAT0_VIN_PGOOD_STAT) == 0x00) {
		*status = CHARGER_STATUS_DISCHARGING;
		return 0;
	}

	ret = i2c_reg_read_byte_dt(&cfg->i2c, BQ25180_ICHG_CTRL, &val);
	if (ret < 0) {
		return ret;
	}

	if ((val & BQ25180_ICHG_CHG_DIS) != 0x00) {
		*status = CHARGER_STATUS_NOT_CHARGING;
		return 0;
	}

	ret = i2c_reg_read_byte_dt(&cfg->i2c, BQ25180_STAT0, &val);
	if (ret < 0) {
		return ret;
	}
```

Sorry one minor change to the order here.

In order we
1. first check PG to see if external power is being provided to the system,. If no power is being provided, we are definitely just discharging. Discharging means we are slowly dying and our battery is our only hope.
2. check if the host (us) has disabled charging, If this is true, we are definitely "not charging". Not charging means power is connected but the battery is not being charged; depending on the system load, we may be dying.
3. Finally go about the rest of the switch case we have below.

Reply:
Ok, I used an extra variable to avoid re-reading stat0 but otherwise it should be what you suggested.